src/ccl_halomod.c

```diff
@@ -0,0 +1,310 @@
+#include "ccl_utils.h"
+#include "math.h"
+#include "stdio.h"
+#include "stdlib.h"
+#include "gsl/gsl_integration.h"
+#include "gsl/gsl_sf_expint.h"
+#include "ccl_background.h"
+#include "ccl_power.h"
+#include "ccl_massfunc.h"
+
+// Cosmology dependence of the virial collapse density according to the spherical-collapse model
+// Fitting function from Bryan & Norman (1998; arXiv:astro-ph/9710107)
+// Here, this is defined relative to the background matter density, not the critical density
+static double Dv_BryanNorman(ccl_cosmology *cosmo, double a, int *status){
+  double Om_mz = ccl_omega_x(cosmo, a, ccl_species_m_label, status);
+  double x = Om_mz-1.;
+  double Dv0 = 18.*pow(M_PI,2);
+  double Dv = (Dv0+82.*x-39.*pow(x,2))/Om_mz;
+  return Dv;
+}
+
+// analytic FT of NFW profile, from Cooray & Sheth (2002; Section 3 of https://arxiv.org/abs/astro-ph/0206508)
+// Normalised such that U(k=0)=1
+static double u_nfw_c(ccl_cosmology *cosmo, double c, double halomass, double k, double a, int *status){
+   
+  double rv, rs, ks, Dv;
+  double f1, f2, f3, fc;
+  double Delta_v = Dv_BryanNorman(cosmo, a, status); // Virial density of haloes
+
+  // Special case to prevent numerical problems if k=0,
+  // the result should be unity here because of the normalisation
+  if(k==0.){    
+    return 1.;    
+  }
+
+  // The general k case
+  else{
+    //Scale radius for NFW (rs=rv/c)
+    rv = ccl_r_delta(cosmo, halomass, a, Delta_v, status);
+    rs = rv/c;
+
+    //Dimensionless wave-number variable
+    ks = k*rs;
+
+    // Various bits for summing together to get final result
+    f1 = sin(ks)*(gsl_sf_Si(ks*(1.+c))-gsl_sf_Si(ks));
+    f2 = cos(ks)*(gsl_sf_Ci(ks*(1.+c))-gsl_sf_Ci(ks));
+    f3 = sin(c*ks)/(ks*(1.+c));
+    fc = log(1.+c)-c/(1.+c);
+  
+    return (f1+f2-f3)/fc;
+  }
+}
+
+// Halo "formation redshift" calculated according to the Bullock et al. (2001) prescription
+// TODO: Actually code this up
+static double z_formation_Bullock(ccl_cosmology *cosmo, double halomass, double a, int *status){ 
+  return 0.;
+}
+
+// The non-linear mass (sigma(M*)=delta_c))
+// TODO: Actually code this up
+static double Mstar(){
+  return 1e13;
+}
+
+// The concentration-mass relation for haloes
+// TODO: make consistency check so that ccl_halo_concentration only runs if called with appropriate definition
+// TODO: e.g. if Delta != 200 rho_{mean}, should not function (or should it?)
+// TODO: should this be moved to the ccl_massfunc.c ?
+double ccl_halo_concentration(ccl_cosmology *cosmo, double halomass, double a, int * status)
+{
+
+  // Set concentration-mass relation
+  // 1 - Bhattaharya et al. (2011)
+  // 2 - Full Bullock et al. (2001)
+  // 3 - Virial Duffy et al. (2008)
+  // 4 - Constant concentration (useful for testing)
+  // 5 - Simple Bullock et al. (2001)
+  int iconc=3;
```

Comment:
This seems to be hard-coded. It would be good to add this as a new option, with an `enum` somewhere to label each option with a keyword (c.f. the `matter_power_spectrum_method` in `ccl_power.h`).

Reply:
Although I wouldn't add this as an option into `ccl_config`, as there don't seem to be any splines/precomputations going on here. It's fine to just add a new argument to the function.

Reply:
Done